sycl/doc/EnvironmentVariables.md

```diff
@@ -28,6 +29,56 @@ compiler and runtime.
 
 `(*) Note: Any means this environment variable is effective when set to any non-null value.`
 
+### `ONEAPI_DEVICE_SELECTOR`
+
+With no environment variables set to say otherwise, all platforms and devices presently on the machine are available. The default choice will be one of them, usually preferring a LevelZero GPU device, if available. The `ONEAPI_DEVICE_SELECTOR` can be used to limit that choice of devices, and to expose sub-devices (aka "tiles") as individual devices.  
+
+A value for this environment value consists of a backend name, followed by a colon ( : ), followed by a device specifier (a named device type, a wildcard symbol ( * ) , or an exact numeric device id).  This can optionally be followed by period ( . )  and then a sub-device specifier, which could be either a wildcard symbol ( * ) or an numeric index (using 0 based counting).   Commas can be used to string multiple device specifiers behind any backend.    Moreover, individual entries can be separated by semi-colons. 
```

Comment:
grammar has been added.

Reply:
> Sub-sub devices (ie two dots) are supported.

What do you mean by "supported"?  Did you verify that using two dots exposes each CCS as a root device in SYCL?  When I talked with @smaslov-intel a few days ago, he said that the PI layer didn't support partitioning a tile (sub-device) into CCS's (sub-sub-devices) yet.  Has that changed?

Reply:
> grammar has been added.

Thanks for adding the grammar.  I was thinking that we could use the grammar instead of the less precise wording, though.  I think our audience will understand BNF grammar.  Many Linux manpages use this grammar to describe the parameters that are accepted by commands.

Could we rewrite the description along these lines?

> With no environment variables set to say otherwise, all platforms and devices presently on the machine are available. The default device selector will choose one of these devices, usually preferring a Level Zero GPU device, if available. The `ONEAPI_DEVICE_SELECTOR` can be used to limit the devices that are available to the application, and it can be used to expose GPU sub-devices (aka "tiles") as SYCL root devices.
>
> The syntax of this environment variable follows this BNF grammar:
>
> ```
> ONEAPI_DEVICE_SELECTOR = <selector-string>
> <selector-string> ::= <term>[;<term>...]
> <term> ::= <backend>:<devices>
> <backend> ::= { * | level_zero | opencl | cuda | hip | esimd_emulator }  // case insensitive
> <devices> ::= <device>[,<device>...]
> <device> ::= { * | cpu | gpu | fpga | <num> | <num>.<num> | <num>.* | *.* }  // case insensitive
> ```
>
> Each term in the grammar selects a collection of devices from a particular backend.  The device names `cpu`, `gpu`, and `fpga` select all devices from that backend with the corresponding type.  A backend's device can also be selected by its numeric index (zero-based) or by using `*` which selects all devices in the backend.
>
> The dot syntax (e.g. `<num>.<num>`) causes one or more GPU sub-devices (aka tiles) to be exposed to the application as SYCL root devices.  For example, `1.0` exposes the first tile of the second device as a SYCL root device.  The syntax `<num>.*` exposes all sub-devices of the give device as SYCL root devices.  The syntax `*.*` exposes all sub-devices of all GPU devices as SYCL root devices.
>
> In general, a term with one or more asterisks (`*`) matches all backends, devices, or sub-devices with the given pattern.  However, a warning is generated if the term does not match anything.  For example, `*:gpu` matches all GPU devices in all backends (ignoring backends with no GPU devices), but it generates a warning if there are no GPU devices in any backend.  Likewise, `level_zero:*.*` matches all sub-devices of partitionable GPUs in the Level Zero backend, but it generates a warning if there are no Level Zero GPU devices that are partitionable into sub-devices.
>
> The device indices are zero-based and are unique only within a backend.  Therefore, `level_zero:0` is a different device from `cuda:0`.  To see the indices of all available devices, run the `sycl-ls` tool.  Note that different backends sometimes expose the same hardware as different "devices".  For example, the `level_zero` and `opencl` backends both expose the Intel GPU devices.
>
> The following examples further illustrate the usage of this environment variable:
>
> | Example | Result |
> -----------|---------
> | `ONEAPI_DEVICE_SELECTOR=opencl:*` | Only the OpenCL devices are available |
> | `ONEAPI_DEVICE_SELECTOR=level_zero:gpu` | Only GPU devices on the Level Zero backend are available.|
> | `ONEAPI_DEVICE_SELECTOR="opencl:gpu;level_zero:gpu"` | GPU devices from both Level Zero and OpenCL are available. (Note that escaping (like quotation marks) will likely be needed when using semi-colon separated entries.) |
> | `ONEAPI_DEVICE_SELECTOR=opencl:gpu,cpu` | Only CPU and GPU devices on the OpenCL backend are available.|
> | `ONEAPI_DEVICE_SELECTOR=opencl:0` | Only the device with index 0 on the OpenCL backend is available. |
> | `ONEAPI_DEVICE_SELECTOR=hip:0,2` | Only devices with indices of 0 and 2 from the HIP backend are available. |
> | `ONEAPI_DEVICE_SELECTOR=opencl:0.*` | All the sub-devices (aka tiles) from the OpenCL device with index 0 are exposed as SYCL root devices.  No other devices are available. |
> | `ONEAPI_DEVICE_SELECTOR=opencl:0.2` | The third tile (2 in zero-based counting) of the OpenCL device with index 0 will be the sole device available.  |
> | `ONEAPI_DEVICE_SELECTOR=level_zero:*,*.*` | Exposes Level Zero devices to the application in two different ways.  Each device (aka "card") is exposed as a SYCL root device and each sub-device (aka "tile") is also exposed as a SYCL root device. |
>
> Notes:
> - The backend argument is always required. An error will be thrown if it is absent.
> - Additionally, the backend MUST be followed by colon ( : ) and at least one device specifier of some sort, else an error is thrown.
> - For sub-devices/tiles, the parent device must support partitioning by `info::partition_property::partition_by_affinity_domain` and `info::partition_affinity_domain::next_partitionable`. See the SYCL 2020 specification for a precise definition.)
> - The semi-colon character ( ; ) is treated specially by many shells, so you may need to enclose the string in quotes if the selection string contains this character. 

I omitted any description of the two-dot syntax here because I'm not sure if it's fully supported.  (See my other comment.)  We can add that back if we decide that it is.

Reply:
@gmlueck  - yes, I tested it on some internal dev hardware and it worked. I was able to partition to sub-devices, and in-turn, partition those to sub-sub-devices, but no further.